text/0029-zero-alloc-messaging-flow.md

```diff
@@ -0,0 +1,390 @@
+
+
+
+![Sending Messages](0029-zero-alloc-messaging-flow/sending.png)
+
+## Receiving Messages
+
+Incoming messages are processed through an incoming message queue, which contains items of this type:
+
+```C#
+internal struct IncomingMessageItem
+{
+    public MessageType MessageType;
+    public FastBufferReader Reader;
+    public ulong SenderId;
+    public NetworkChannel ReceivingChannel;
+    public float Timestamp;
+}
+```
+
+A queue of this type exists as a simple array with a configurable fixed size. If it gets full, we begin discarding messages to avoid allocating to resize the array.
+
+At the point of receiving, we construct a new `NativeArray<byte>` using `Allocator.TempJob` and copy the received data into it, then create a `FastBufferReader` from that NativeArray. The `BatchHeader` is read to determine the total size of the message batch. We then iterate, reading messages (but not processing them) and appending them to the `InternalMessageQueue`, using the `MessageHeader`'s `MessageSize` value to find the beginning of each new message, until we have read the full batch.
+
+We then iterate through `IncomingMessageQueue` as a separate step (thus picking up loopback messages from the previous frame first), looking up each registered `MessageHandler` via a fixed array indexed on `MessageType`, and call the handler without any knowledge of the message it will be processing.
+
+Finally, an additional queue exists: The dispose queue. Like the incoming message queues, this queue is a fixed-size array of `NativeArray<byte>` objects sized to the combined size of all update queues. Once each buffer has been processed, it's added to the dispose queue, where it will be disposed at the end of the frame after all messages pointing into it have been handled. Disposing is deferred to the end of the frame in order to ensure that the buffers stay alive when `DelayUtility` is used.
+
+![Receiving Messages](0029-zero-alloc-messaging-flow/receiving.png)
+
+## Custom Messages
+
+Custom messages as they are, using buffers, no longer need to exist. Since the message handling code is part of the message definition, users have the freedom to define their own messages. Our internal code doesn't need to know anything about the messages themselves in order to send or receive them, so users can create any custom message they want the same way we define our own internal messages, and send them the same way we send ours, and their message handler will be magically invoked when the message is received. The only gotcha is the MessageType enum - since users can't extend the enum itself, we provide a `MessageType.Custom` that's always at the end of the enum, and users can operate as such:
```

Comment:
Users having the ability to create their own message types with associated IMessage implementation is very cool indeed!